src/azul/terraform.py

```diff
@@ -721,6 +722,29 @@ def tf_config(self, app_name):
                 key.replace('.', '_'): value
                 for key, value in bucket_notifications.items()
             }
+            # To prevent a race condition by Terraform, we make the bucket
+            # notifications depend on the related aws_lambda_permission and
+            # aws_lambda_function resources.
+            for notification in resources['aws_s3_bucket_notification'].values():
+                lambda_arns = [
+                    lambda_function['lambda_function_arn']
+                    for lambda_function in notification['lambda_function']
+                ]
+                lambda_resources = []
+                permission_resources = []
+                for key, value in resources['aws_lambda_permission'].items():
+                    lambda_arn = value['function_name']
+                    if lambda_arn in lambda_arns:
+                        match = re.fullmatch(r'\${(.*)\.arn}', lambda_arn)
+                        assert match is not None, lambda_arn
+                        lambda_resources.append(match.group(1))
+                        permission_resources.append(f'aws_lambda_permission.{key}')
+                assert len(lambda_arns) == len(lambda_resources), (lambda_arns, lambda_resources)
+                # Since the aws_s3_bucket_notification resource likely won't have a
+                # depends_on argument, we conditionally add it before filling it up.
+                depends_on = notification.setdefault('depends_on', [])
```

Comment:
It would be surprising if there already were a `depends_on` property. And if there were one, you'd have to worry about not creating duplicate entries. So as usual, instead of making a sloppy attempt of accommodating that case, you should be strict and fail. Then you *know* that what you wrote is correct, you made the assumptions clear, and you won't have to worry about covering a scenario you have never actually observed.